features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/forward/ForwardMessagesPresenter.kt

```diff
@@ -0,0 +1,135 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.element.android.features.messages.impl.forward
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.collectAsState
+import androidx.compose.runtime.derivedStateOf
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.runtime.setValue
+import dagger.assisted.Assisted
+import dagger.assisted.AssistedFactory
+import dagger.assisted.AssistedInject
+import io.element.android.libraries.architecture.Async
+import io.element.android.libraries.architecture.Presenter
+import io.element.android.libraries.architecture.isLoading
+import io.element.android.libraries.designsystem.theme.components.SearchBarResultState
+import io.element.android.libraries.matrix.api.MatrixClient
+import io.element.android.libraries.matrix.api.core.EventId
+import io.element.android.libraries.matrix.api.core.RoomId
+import io.element.android.libraries.matrix.api.room.MatrixRoom
+import io.element.android.libraries.matrix.api.room.RoomSummary
+import io.element.android.libraries.matrix.api.room.RoomSummaryDetails
+import kotlinx.collections.immutable.ImmutableList
+import kotlinx.collections.immutable.persistentListOf
+import kotlinx.collections.immutable.toPersistentList
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+class ForwardMessagesPresenter @AssistedInject constructor(
+    @Assisted eventId: String,
+    private val room: MatrixRoom,
+    private val matrixCoroutineScope: CoroutineScope,
+    private val client: MatrixClient,
+) : Presenter<ForwardMessagesState> {
+
+    private val eventId: EventId = EventId(eventId)
+
+    @AssistedFactory
+    interface Factory {
+        fun create(eventId: String): ForwardMessagesPresenter
+    }
+
+    @Composable
+    override fun present(): ForwardMessagesState {
+        var selectedRooms by remember { mutableStateOf(persistentListOf<RoomSummaryDetails>()) }
+        var query by remember { mutableStateOf<String>("") }
+        var isSearchActive by remember { mutableStateOf(false) }
+        var results: SearchBarResultState<ImmutableList<RoomSummaryDetails>> by remember { mutableStateOf(SearchBarResultState.NotSearching()) }
+        val forwardingActionState: MutableState<Async<ImmutableList<RoomId>>> = remember { mutableStateOf(Async.Uninitialized) }
+
+        val summaries by client.roomSummaryDataSource.roomSummaries().collectAsState()
+
+        LaunchedEffect(query, summaries) {
+            val filteredSummaries = summaries.filterIsInstance<RoomSummary.Filled>()
```

Comment:
We should also filter out:

1. Duplicates, as duplicate IDs will crash the LazyColumn, and we're not guaranteed there won't be any.
2. Rooms we cannot write to (take powerlevel into account, if possible).

Reply:
We shouldn't, but sadly we do. It seems like it happens on iOS too. When I asked @manuroe he mentioned they still have this issue with RoomList.

Reply:
Also, it seems like:

> Rooms we cannot write to (take powerlevel into account, if possible).

It's quite difficult to do right now with no performance hits. As far as I could see there is no option in slidingsync, sor for each item received in `RoomSummaryDataSource` we'd have to get its full room, the current member, and then its powerlevel/`canSend(eventType)`.